### ISSUE TITLE
About doing transfer learning

### ISSUE THREAD

**Issue by vincentwen1995** [ildoonet/tf-pose-estimation#162]
Hi. Currently I am trying to combine the OpenPose network with a pre-trained network (modified ResNet-50) and do some transfer learning. According to the original OpenPose paper, the first 10 layers of VGG19 is used as the feature extractor, so I also located a specific layer in the pre-trained network whose output is equivalently sufficient for the same level of feature extracting as the VGG19. 
My implementation is to first create a decorated layer which takes the loaded frozen graph (stored in .pbtxt) as an argument and connect the input tensor of the graph with the layer's input tensor (image after normalization to [0, 1)) and return the output tensor of the graph, as shown below in the code.

```
@layer
def BlackBox(self, input, graph_def, name):
    g = tf.get_default_graph()
    with g.as_default():
        out = tf.import_graph_def(
            graph_def=graph_def,
            input_map={'input:0': input},
            return_elements=[
                'specific_layer/Relu'],
        )
        return out[0].outputs[0]
```

Then create a new subclass of the BaseNetwork class that has the setup method with first loading the frozen graph from the .pbtxt and then connecting each layer, as shown below in the code.

```
def setup(self):
    graph_def = graph_pb2.GraphDef()
    with open('/path-to-graph/graph.pbtxt', 'rb') as f:
        text_format.Merge(f.read(), graph_def)

        (self.feed('image')
             .normalize_BlackBox(name='preprocess')
             .BlackBox(graph_def=graph_def, name='BlackBox')
             .conv(3, 3, 256, 1, 1, name='conv4_3_CPM')
             .conv(3, 3, 128, 1, 1, name='conv4_4_CPM')          # *****
        ......
```

However, the above implementation gives an error during training, stating that the convolutional layer following the BlackBox layer cannot acquire a valid shape of the input, which is the output tensor returned by the BlackBox layer.
As I added some debugging info to the code, I noticed that the shape of the placeholder that represents the input image is successfully passed until the input tensor of the BlackBox layer, but the shape of the output tensor of the BlackBox layer is not shown. Thus, I suppose the problem could be with the fact the the shape of the placholder is not passed through the loaded graph.

So my question is what would be the correct way of implementation and what did I get wrong about the architecture of the code?

UPDATE:
As I researched further into it, I found that the problem is with the fact that the shape information of the tensors were not saved into the graphdef, which I "solved" with reshaping the output tensor to the shape it should be in the network. However, I am still not sure whether this is the correct way to implement transfer learning on a frozen graph, so I am still open to suggestions.

As I proceeded further with my method, I was able to run the training.py before I received errors about the validation. According to the log, the frozen graph is successfully connected to the training network. However the losses are NaN 's, so I suppose there are still some connections are incorrect.

**Comment by ouceduxzk** [comment]
A frozen graph is not prepared for trainining, it is prepare for inference, instead of doing transfer learning from frozen graph, you use the checkpoint like in this repo did, use the pretrained mobilenet_whatever on imagenet to train a pose estimator

**Comment by vincentwen1995** [comment]
@ouceduxzk Aren't those the same technically in this case? Since only the output of a certain-level feature extractor is needed for training on the PAF and Confidence Map, I use the frozen graph for inference and extract the output of that certain layer and then feed it into the pose training network. 
As I proceeded further with my method, I was able to run the training.py before I received errors about the validation. According to the log, the frozen graph is successfully connected to the training network. However the losses are NaN 's, so I suppose there are still some connections are incorrect.
